### PR TITLE
Easing into the 'pre-commit' framework (pcf)

### PR DESCRIPTION
**Pull request by safl**
This is a re-work of @Baekalfen initiative to introduce ``pre-commit`` at https://github.com/OpenMPDK/xNVMe/pull/48
This picks up the commits and does the following:
* Remove a handful of scripts and tools that are out-dated
* Do the clang-format using the existing clang-format definitions and adding a script using them
* Add helpers to ``Makefile`` invoking ``pre-commit`` manually
* Add ``pre-commit`` to ``.githooks/pre-commit`` with
* Add the different pre-commit hooks/plugins
* Apply them and manually fix-up lint issues in a bulk-commit
* Add the commit to git-blame-ignore
* Add a 'pre-commit' badge in README

### CODE REVIEW

File: scripts/pcf_clang_format.py [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+    args.style_file = expand_path(args.style_file)
+
+    return args
+
+
+def clang_format(args):
+    """Invoke clang-format"""
+
+    # Form arguments to clang-format parameter "-style"
+    style = []
```
Comment (Baekalfen) [inline scripts/pcf_clang_format.py:51]: Unnecessary definition.

File: scripts/pcf_clang_format.py [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+    with open(args.style_file) as sfd:
```
Comment (Baekalfen) [inline scripts/pcf_clang_format.py:52]: Unpythonic variable naming: `sfd` -> `style_file`, `f` or like that

**Review by Baekalfen — COMMENTED** [review]
Tiny comments, will work regardless

**Review by Baekalfen — COMMENTED** [review]
In commit message:

> With the option of running it on everything:
  pre-commit run clang-format --all

`--all` refers to all hooks, `--all-files` will make it run on unstaged files as well.

File: Makefile [hunk 3]
```diff
@@ -122,6 +122,24 @@ git-setup:
...
+	@pre-commit run
+	@echo "## xNVME: format [DONE]"
+
+define format-all-help
+# run code format (style, code-conventions and language-integrity) on staged and committed changes
+endef
+.PHONY: format-all
+format-all:
+	@echo "## xNVMe: format-all"
+	@pre-commit run --all
```
Comment (Baekalfen) [inline Makefile:140]: `--all` refers to all hooks, `--all-files` will make it run on unstaged files as well.

File: .githooks/pre-commit [hunk 4]
```diff
@@ -1,25 +1,45 @@
...
 ==============================================================================
- Error: invalid source-format
+Error: the 'pre-commit' framework https://pre-commit.com/ is not available
 ==============================================================================
 
-* You can inspect the current format-issues here: 'scripts/source-format.log'
+You can install it via the Python Package Index:
 
-* Run the source-formatter / fix the source-format issues
+  pip3 install --user pre-commit
```
Comment (Baekalfen) [inline .githooks/pre-commit:16]: Additional step for commit message check:
```
pre-commit install --hook-type commit-msg
```

File: .pre-commit-config.yaml [hunk 5]
```diff
@@ -59,3 +59,69 @@ repos:
...
+    - '--extend-ignore=E203' # ignore "whitespace before ':'"
+- repo: https://github.com/jendrikseipp/vulture
+  rev: 'v2.3'
+  hooks:
+  - id: vulture
+    name: Python-lint-vulture
+    types: [python]
+    pass_filenames: true
+    # excluding sphinx-conf as it would break to change it
+    exclude: docs/autogen/conf.py$
```
Comment (Baekalfen) [inline .pre-commit-config.yaml:96]: I think we need a bit on config to make it actually trigger. But it doesn't break anything to have it in.

File: .github/workflows/verify.yml [hunk 6]
```diff
@@ -56,49 +56,24 @@ jobs:
...
-        apt -qy install wget gnupg2 lsb-release software-properties-common
-        wget -O - https://apt.llvm.org/llvm-snapshot.gpg.key | apt-key add -
-        add-apt-repository -y "deb http://apt.llvm.org/bullseye/ llvm-toolchain-bullseye-14 main"
-        apt -qy update
-        apt -qy install clang-format-14
-        update-alternatives --install /usr/bin/clang-format clang-format /usr/bin/clang-format-14 100
-
-    - name: xNVMe, run the source-format script
+        apt-get -qy update && apt-get -qy install shellcheck
+        pip3 install virtualenv==20.0.33
```
Comment (Baekalfen) [inline .github/workflows/verify.yml:73]: Shouldn't this be in `refenv/xnvme-devtools-ci:latest`?

File: .github/workflows/verify.yml [hunk 7]
```diff
@@ -56,49 +56,24 @@ jobs:
...
-      if: failure()
-      run: |
-        cat scripts/source-format.log
-
-    - name: Upload source-format log
-      uses: actions/upload-artifact@v2
-      with:
-        name: source-format-issues
-        path: scripts/source-format.log
+        pre-commit run --all
```
Comment (Baekalfen) [inline .github/workflows/verify.yml:76]: I think you'll need `--all-files` https://github.com/pre-commit/action/blob/main/action.yml

**Comment by Baekalfen** [comment]
Maybe I'm wrong about the `--all` vs `--all-files` arg. I can't actually find `--all` in the docs.

File: .github/workflows/verify.yml (hunk 6 above)
Comment (safl) [reply .github/workflows/verify.yml:73]: Nice, catch, yes. Just added it here until it "worked". Fixing...

File: .github/workflows/verify.yml (hunk 7 above)
Comment (safl) [reply .github/workflows/verify.yml:76]: Nice catch. Curiously enough it "works" with ``--all`` but that must be a fluke-accident from how the parser works... fixing...

File: scripts/pcf_clang_format.py (hunk 2 above)
Comment (safl) [reply scripts/pcf_clang_format.py:52]: Yeah, I see your point. My C abbreviation habits leaking in here. Fixing it.

**Comment by safl** [comment]
> 

I think you are right, however, for some reason ``--all`` behaves the same as ``--all-files``. Changing it to ``--all-files`` since that is what the ``--help`` says..

File: .githooks/pre-commit [hunk 8]
```diff
@@ -1,25 +1,45 @@
...
+when using Bash:
+
+  export GIT_HOOKS_OFF=1
+
+==============================================================================
+EOF
+	exit 1
+}
+
+if ! pre-commit run; then
```
[the remark this answers is not in the document]
Comment (safl) [reply .githooks/pre-commit:31]: Hmm, yeah... this overlaps a bit with the whether or not ``GIT_HOOKS_OFF=1`` is actually needed at all.
I just kind of like that the script tells the user what to do and how to easily work around it.
But dropping the custom-script with ``GIT_HOOKS_OFF`` and relying on ``pre-commit install`` would ensure a proper invocation of ``pre-commit``. Hmm...

File: .githooks/pre-commit (hunk 4 above)
Comment (safl) [reply .githooks/pre-commit:16]: Doesn't ``pre-commit`` auto-install these on first-run? Along with all the others?

File: Makefile (hunk 3 above)
Comment (safl) [reply Makefile]: As discussed elsewhere, it is curious... but it is not changed to ``--all-files``.

File: scripts/pcf_clang_format.py (hunk 2 above)
Comment (safl) [reply scripts/pcf_clang_format.py]: Fixed.

File: Makefile (hunk 3 above)
Comment (safl) [reply Makefile]: Fixed.

**Comment by safl** [comment]
Thanks for the input @Baekalfen ! I have addressed the comments, except for one/two things:

- [ ] Question: Is the addition of ``pre-commit install --hook-type commit-msg`` required? Won't pre-commit take care of that along with along the other plugin/hook-installations?
- [ ] Whether or not ``pre-commit install`` should take of ``.githooks/pre-commit`` or whether the current approach is good enough. There is a clear argument that using ``pre-commit install`` would be preferred as it sets up ``pre-commit`` according the local environment, and that one should just use ``--no-verify`` when committing/rebasing in case one wants to ignore ``pre-commit``. Let me just mull on this one...

File: .pre-commit-config.yaml [hunk 9]
```diff
@@ -0,0 +1,135 @@
...
+    types: [yaml]
+
+# Python: for all the helpers in scripts/*
+- repo: https://github.com/psf/black
+  rev: '22.1.0'
+  hooks:
+  - id: black
+    name: Python-format-black
+    args:
+    - '--line-length=79'
```
Comment (Baekalfen) [inline .pre-commit-config.yaml:70]: Can't say I like the 79 character limit. That's going to be in the way

File: .pre-commit-config.yaml [hunk 10]
```diff
@@ -0,0 +1,135 @@
...
+    - '--mapping=2'
+    - '--sequence=2'
+    - '--offset=0'
+    - '--width=120'
+    - '--preserve-quotes'
+    types: [yaml]
+
+# Python: for all the helpers in scripts/*
+- repo: https://github.com/psf/black
+  rev: '22.1.0'
```
Comment (Baekalfen) [inline .pre-commit-config.yaml:65]: Should be updated to `'22.3.0'`. This fixes a bug which is blocking CI right now

Comment (Baekalfen) [reply .pre-commit-config.yaml:65]: https://github.com/psf/black/issues/2964

File: .githooks/pre-commit (hunk 4 above)
Comment (Baekalfen) [reply .githooks/pre-commit]: Could be it was only explicitly needed, as I was adding the commit-msg hook after the regular install.

But if we have our own pre-commit script, we need a hook in Git to trigger on commit-msg.

File: .pre-commit-config.yaml (hunk 5 above)
Comment (Baekalfen) [reply .pre-commit-config.yaml:96]: Passing filenames will presumably make `vulture` only check for dead-code within each file. You need to supply directories/packages to check inter-dependability.

File: .githooks/pre-commit (hunk 8 above)
[the remark this answers is not in the document]
Comment (safl) [reply .githooks/pre-commit]: Adjusted according to offline discussion.

File: .pre-commit-config.yaml (hunk 10 above)
Comment (safl) [reply .pre-commit-config.yaml]: Fixed

File: .pre-commit-config.yaml (hunk 9 above)
Comment (safl) [reply .pre-commit-config.yaml]: Bumped up to use the default provided by black.

File: docs/autogen/apigen.py [hunk 11]
```diff
@@ -3,44 +3,48 @@
...
+
 import jinja2
 
 DECLARATIONS = {
     "func": [],
     "macro": [],
     "struct": [],
     "enum": [],
-}
+}  # type: dict
```
Comment (Baekalfen) [inline docs/autogen/apigen.py:21]: Is it `mypy` that requires this? Why not:
```python
DECLARATIONS : dict = {
     "func": [],
     "macro": [],
     "struct": [],
     "enum": [],
 }
```

File: docs/autogen/apigen.py [hunk 12]
```diff
@@ -157,24 +160,25 @@ def main(args):
...
+        "xnvme",
     ]
 
     syms = symbols(args, namespaces)
 
     pps = {}
-    for ns in syms:
-        pps[ns] = find_pp(args, syms[ns])
+    for nsprefix, val in syms.items():
+        pps[nsprefix] = find_pp(val)
```
Comment (Baekalfen) [inline docs/autogen/apigen.py:170]: Don't fix this unless you want to, but it's perfect for a dict-comprehension:
```python
pps = {k:find_pp(val) for k,v in syms.items()}
```

File: .pre-commit-config.yaml [hunk 13]
```diff
@@ -0,0 +1,134 @@
...
+# pre-commit runs pylint from an isolated virtualenv. Many of pylint's checks perform dynamic
+# analysis which will fail there. You may find configuring pylint as a local hook more useful.
+- repo: https://github.com/pre-commit/mirrors-pylint
+  rev: 'v3.0.0a4'
+  hooks:
+  - id: pylint
+    name: Python-lint-pylint
+    args:
+    - '--jobs=4' # Run with 4 threads
+    - '--good-names=i,j,k,n,f,_,Run,ex' # Add 'f' for with open()
```
Comment (Baekalfen) [inline .pre-commit-config.yaml:116]: `v` is also common in `for k,v in some_dict.items()`

**Review by Baekalfen — APPROVED** [review]
Looks good. Nothing critical

File: docs/autogen/apigen.py (hunk 11 above)
Comment (safl) [reply docs/autogen/apigen.py:21]: Yup, mypy complains when the type-annotation is not there

File: .pre-commit-config.yaml (hunk 13 above)
Comment (safl) [reply .pre-commit-config.yaml:116]: Fixing...

Comment (safl) [reply .pre-commit-config.yaml]: Fixed.

**Comment by safl** [comment]
Thanks for the feedback. When checks pass I'll move it forward.

**Pull request merged**